Frontend asset loading

rtibbles opened the pull request:
## Summary

This is a WIP to implement and provide an example hook for `base.html` of both synchronous and asynchronous frontend asset loading.

## TODO

- [x] Have tests been written for the new code?
- [x] Has documentation been written/updated?
- [X] New dependencies (if any) added to requirements file
- [x] Add an entry to CHANGELOG.rst
- [x] Add yourself it AUTHORS.rst if you don't appear there

## Reviewer guidance

This PR also adds in JS Linting during the webpack build process, in order to compensate for the fact that the JSHint precommit hook doesn't work on Windows and had to be removed.

In addition, it simplifies the naming of frontend assets to provide a simpler naming convention, based solely on the name of their Python module in Kolibri, and the name of the Class in which they are defined in `kolibri_plugins.py`.

Finally, it provides a Javascript asynchronous frontend asset loader that can load both CSS and Javascript, in order to allow plugins to be registered to the frontend asynchronously and loaded when needed.

Adds a function for creating template tags with a hook to allow arbitrary plugins to plugin synchronously and asynchronously into any template.

Creates tags for these for base.html to allow for extension of base.html frontend functionality and as an example.

## Issues addressed

Fixes #41, and fixes #38 

## Documentation

Much inline documentation. Also updates frontend.rst.

rtibbles commented:
Woops, broke coverage.

rtibbles commented:
Woot, the asset-loader module has *full* code coverage!

aronasorman commented on .jshintrc:
```diff
@@ -0,0 +1,20 @@
+{
```
Is it possible to put a comment with a link pointing to where these flags are defined?

rtibbles replied:
Do we have one in travis.yml, or .gitignore? It's in the JSHint documentation.

aronasorman commented on assets/src/logging.js:
```diff
@@ -1,3 +1,4 @@
+'use strict';
```
If everything is already linted as strict, why do we need to still add this header?

rtibbles replied:
`'use strict';` is an environment directive that is parsed by the browser itself. It will enforce the code to be executed in a strict context, which means that the Javascript engine will be less forgiving.

See here for details: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode

rtibbles commented:
Pushing latest changes in case anyone wants to look at them.

Tests have not been updated, so will break. Documentation needs to be written.

rtibbles commented:
Some tests updated now, have also made a new tox_env for node, so that we now only run JS tests once!

rtibbles commented:
@aronasorman This is now ready for final review.

rtibbles commented:
![gifrecord_2016-03-14_180602](https://cloud.githubusercontent.com/assets/1680573/13764710/9032dd2e-ea0f-11e5-8230-164637473025.gif)

Demonstration of callback buffering and arg passing in async loading mode!

aronasorman commented:
Merge conflict.

aronasorman commented on kolibri/core/assets/src/mediator/mediator.js:
```diff
@@ -1,7 +1,127 @@
+        }
+    },
+
+    register_plugin_async: function(plugin_name, plugin_urls, events) {
```
`register_plugin_async` seems to have been defined twice. Is that intentional?

rtibbles replied:
I don't see it - are you confusing it with `register_plugin_sync`?

aronasorman commented on kolibri/core/assets/src/kolibri_core_app.js:
```diff
@@ -1,8 +1,32 @@
+var Kolibri = _.extend({
+    lib: lib,
+    plugins: Mediator._plugin_registry
+}, _.pick(Mediator, public_methods));
```
I'm not too familiar with `lodash`. What does this entire `_.extend` and `_.pick` snippet do?

rtibbles replied:
`_.pick` takes certain attributes (the keys in `public_methods`) from the Mediator object and then `_.extend` adds them to the object that has the `lib` and `plugins` attributes defined.

aronasorman commented on webpack.config.js:
```diff
@@ -11,6 +11,18 @@ var js_source_dirs = [
+        'lodash': kolibri_name + '.lib._',
+        'backbone': kolibri_name + '.lib.Backbone',
+        'loglevel': kolibri_name + '.lib.loglevel',
+        'kolibri': kolibri_name
```
So is this where the namespacing to `lib` happens? It seems like we return a dict in here, how exactly do these libraries then get namespaced when loading in the browser?

rtibbles replied:
This is a convenience mapping to ensure that no plugins bundle in lodash, backbone, or loglevel if they do a direct require call on them. Instead, the reference will be converted into a reference to the copy that is on the core Kolibri object.

aronasorman commented on kolibri/plugins/example_plugin/assets/example/example_plugin.js:
```diff
@@ -1,5 +1,24 @@
+        'something_happened': 'hello_world'
+    },
+    hello_world: function(plugin) {
+        if (plugin === this) {
```
Why is this `if` statement required? So this callback is not guaranteed to be called on this plugin's context?

rtibbles replied:
Originally, I was triggering this from the 'plugin_registered' event, which also emits the plugin object as the first argument passed to a callback, this was just checking that the plugin in question was registered.

Ultimately, we can parse arbitrary data and arguments into callback functions from events.

aronasorman commented:
@rtibbles can you point to where the whole event-callback machinery is implemented? I'd like to know when callbacks are triggered, and what the surounding context is (this is prompted after seeing that callbacks can be called outside of their plugin context.)

aronasorman commented on docs/dev/frontend.rst:
```diff
@@ -28,30 +28,47 @@ As such for a particular component of a frontend plugin, would appear in the ass
+        },
+
+        goodbye_world: function(message) {
+            logging.info('Goodbye, cruel world!', message);
```
Such is the life of a plugin.

aronasorman commented:
Can you add a section in `frontend.rst` showing which libraries they have access to and don't have to download (a.k.a. `Kolibri.libs.*`)

aronasorman commented on kolibri/plugins/base.py:
```diff
@@ -111,23 +111,33 @@ def webpack_bundle_data(cls):
         "entry_file" - is the Javascript file that defines the plugin.
-        "external" - an optional flag used only by the kolibri_core plugin.
+        "external" - an optional flag currently used only by the core plugin.
+        "core" - an optional flag *only* ever used by the core plugin.
```
Can you also document what the effects are of having the `core` flag?

rtibbles replied:
Wilco.

MCGallaspy commented:
Does `README.rst` need to be updated (namely the instructions for running the dev server)?

MCGallaspy commented:
Also, which version of node do we support/recommend?

MCGallaspy commented:
I was trying this out myself. This commit causes an error for me when running `npm run watch`: https://github.com/rtibbles/kolibri/commit/2f3129fd556882a29b64fbf479ad7b66014a73a8

This works fine, however:

```js
var onloadcss = require('fg-loadcss/src/onloadCSS');
/*
...
*/
module.exports = {
    assetLoader: assetLoader,
    onloadCSS: onloadcss
};
```

rtibbles commented:
And does nothing, as `onloadCSS.js` fails to define a module.exports. The exports-loader that is causing the error for you allows the global variable defined in that file to be explicitly exported and used.

I assume you have run `npm install`, so I guess there may be an issue with the export-loader and windows? None reported here - https://github.com/webpack/exports-loader/issues

rtibbles commented:
@MCGallaspy Seems to be an issue with node v5 (I have replicated with node v5 locally).

Let's pin our node at v4 for now.

rtibbles commented:
Hrm, seems this is not true. Having cleaned out my node_modules dir, I am able to replicate this on all versions of node. Most curious.

rtibbles commented:
OK, got the external loader working properly on Node v4 and Node v5.

rtibbles commented:
@MCGallaspy README is already updated with how to run the npm scripts.

Should be agnostic to node version.

rtibbles commented:
@aronasorman Note: All callbacks that are registered are executed in a way that ensures that their `this` in the execution context refers to the plugin that registered that callback. That way, every callback has a guaranteed execution context. See above for my explanation of the check that I made - it was not checking the value of `this` but rather checking the value of the event argument.

MCGallaspy commented:
> Merge branch 'master' into frontend_asset_loading …

Personally I've come to prefer rebasing, since it results in cleaner commit histories. Of course, it does become a little problematic for very large or long-standing PRs...

MCGallaspy commented:
> README is already updated with how to run the npm scripts.

I see that now, but it's under the "Testing" heading, which is a little odd. I'd expect to see it under "Running Kolibri", and it would be better still if I could run a dev server with one command instead of two.

aronasorman commented on kolibri/core/assets/src/mediator/mediator.js:
```diff
@@ -1,3 +1,241 @@
+    this._register_one_time_events(plugin);
+
+    this._plugin_registry[plugin.name] = plugin;
+    this._clear_async_callbacks(plugin);
```
We should document that async callbacks are cleared after plugin loading.

aronasorman commented on kolibri/core/assets/src/mediator/mediator.js:
```diff
@@ -1,3 +1,241 @@
+
+    this._plugin_registry[plugin.name] = plugin;
+    this._clear_async_callbacks(plugin);
+    this._execute_callback_buffer(plugin);
```
Memory leaks are a concern. Not too much of a problem so early on, but let's watch out for this in the future.

DXCanas commented:
Running into some weird issues on production server, discussed with Richard on 3/28. Traceback here:

`Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/

Django Version: 1.9.1
Python Version: 3.4.3
Installed Applications:
['kolibri.core',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'kolibri.auth.apps.KolibriAuthConfig',
 'kolibri.content']
Installed Middleware:
('django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware')



Traceback:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in __getitem__
  86.             return self._engines[alias]

During handling of the above exception ('django'), another exception occurred:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_package_libraries
  144.             module = import_module(entry[1])

File "/home/david/Envs/kolibri/lib/python3.4/importlib/__init__.py" in import_module
  109.     return _bootstrap._gcd_import(name[level:], package, level)

File "/home/david/Projects/kolibri/kolibri/core/templatetags/test/test_kolibri_tags.py" in <module>
  4. from mock.mock import patch

During handling of the above exception (No module named 'mock'), another exception occurred:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/core/handlers/base.py" in get_response
  174.                     response = self.process_exception_by_middleware(e, request)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/core/handlers/base.py" in get_response
  172.                     response = response.render()

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in render
  160.             self.content = self.rendered_content

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in rendered_content
  135.         template = self._resolve_template(self.template_name)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in _resolve_template
  90.         new_template = self.resolve_template(template)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in resolve_template
  80.             return select_template(template, using=self.using)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/loader.py" in select_template
  55.     engines = _engine_list(using)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/loader.py" in _engine_list
  143.     return engines.all() if using is None else [engines[using]]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in all
  110.         return [self[alias] for alias in self]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in <listcomp>
  110.         return [self[alias] for alias in self]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in __getitem__
  101.             engine = engine_cls(params)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in __init__
  31.         options['libraries'] = self.get_templatetag_libraries(libraries)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_templatetag_libraries
  49.         libraries = get_installed_libraries()

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_installed_libraries
  131.             for name in get_package_libraries(pkg):

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_package_libraries
  148.                 "trying to load '%s': %s" % (entry[1], e)

Exception Type: InvalidTemplateLibrary at /
Exception Value: Invalid template library specified. ImportError raised when trying to load 'kolibri.core.templatetags.test.test_kolibri_tags': No module named 'mock'
`

MCGallaspy commented:
I think with python 3, the `mock` module is actually `unittest.mock`. Looks like cause of the import error here in `test_kolibri_tags.py`. The fix is to use a conditional import like this:

```python
try:
    import mock  # Python 2
except ImportError:
    import unittest.mock as mock  # Python 3
```

rtibbles commented:
No, this works fine in Python 3, as we have mock in the test requirements.
The issue is that Django is being more magic than I thought and importing
the test files as they've are inside the template tags folder. I will move
the tests out.

On Tue, 29 Mar 2016, 09:52 Michael Gallaspy, <notifications@github.com>
wrote:

> I think with python 3, the mock module is actually unittest.mock. Looks
> like cause of the import error here in test_kolibri_tags.py. The fix is
> to use a conditional import like this:
>
> try:
>     import mock  # Python 2except ImportError:
>     import unittest.mock as mock  # Python 3
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/learningequality/kolibri/pull/49#issuecomment-202995361>
>

MCGallaspy commented:
Oh, I see. Perhaps the tests directory should not be a module? I'm not really sure what the convention is here.

rtibbles commented:
Two regression tests needed here: both for checking that `this` is set correctly. Will write before they are merged.

rtibbles commented:
Have added appropriate regression tests. Fixed previous test failure.

Edited README.rst to move things out of testing that don't belong.

Fixed the testing issue for template tags.

Added line by line comments to some of the more obscure code.

indirectlylit commented on .gitignore:
```diff
@@ -53,5 +54,5 @@ docs/modules.rst
 # webpack built files
-*stats.json
+kolibri/**/build/*
 kolibri/**/static/*
```
I understand why this is necessary here, but the wildcard might cause brief confusion if someone decides to call a directory 'static' eventually

indirectlylit commented on assets/src/parse_bundle_plugin.js:
```diff
@@ -8,14 +9,15 @@ var BundleTracker = require('webpack-bundle-tracker');
 var path = require("path");
 var logging = require('./logging');
+var webpack = require('webpack');
+var base_config = require('../../webpack.config.base');
```
relative path - code smell?

rtibbles replied:
Trying to avoid relative paths as much as possible (hence the 'component' structure I have urged in the Frontend Docs - where code that will be imported together should be in the same directory) but in this case, I think a relative path is unavoidable unless we define an alias for node somewhere else.

rtibbles commented:
After some discussion with @indirectlylit and @MCGallaspy, and due to some issues with the approach taken to inspect the Javascript code and extract the `events` and `once` hashes at compilation time to allow for asynchronous loading - I have now changed the `events` and `once` hashes to be defined within the classes subclassed from `KolibriFrontEndBase`.

Configuration of these hashes now happens from within `kolibri_plugin.py`, and is then injected into the Javascript code at build time.

MCGallaspy commented:
Can you also update this section: https://github.com/rtibbles/kolibri/blob/frontend_asset_loading/docs/dev/frontend.rst#writing-frontend-plugins

Also, who is responsible for merging this PR?

rtibbles commented:
Good catch. Changing.

rtibbles commented:
@aronasorman is assigned - I'll see if he is happy to sign it off.

indirectlylit commented on kolibri/plugins/assets/src/plugin_base/plugin_base.js:
```diff
@@ -0,0 +1,33 @@
+// The Backbone Model extend method is a standalone function that is used to extend many Backbone objects.
+// We use it here in preference to rolling our own to allow for extension of Plugins.
+
+Plugin.extend = Backbone.Model.extend;
```
Are plug-ins backbone models? If not, does this introduce a dependency on backbone just for this method?

indirectlylit replied:
answering my own Q: Looks like `KolibriModule` is not a backbone model. We're also using `Backbone.Events` elsewhere.

indirectlylit replied:
IMO it would be nice to inline the `extend` function to remove this instance of the Backbone dependency.

rtibbles replied:
Probably worthwhile.

indirectlylit commented on kolibri/core/kolibri_plugin.py:
```diff
@@ -38,6 +38,7 @@ class KolibriCoreFrontEnd(KolibriFrontEndPluginBase):
     """
     entry_file = "assets/src/kolibri_core_app.js"
     external = True
+    core = True
```
Why is this added to `class KolibriCoreFrontEnd(KolibriFrontEndPluginBase)` but not `class KolibriCore(KolibriPluginBase)`?

indirectlylit replied:
(Looks like `KolibriCore` is never used, and currently contains what looks like an example subclass implementation?)

rtibbles replied:
This 'core' property is only for defining a frontend KolibriModule as core. Should mention in the base class docs - if not, I'll add.

indirectlylit commented:
I'm a bit confused by the various webpack config files in the root directory:

* `parse_bundle_plugin.js` and `karma.conf.js` reference `webpack.config.base.js`
* `webpack.config.dev.js` references `webpack.config.js`
* `webpack.config.base.js` and `webpack.config.js` _do not_ reference each other

Seems like perhaps both `webpack.config.base.js` and `webpack.config.js` are trying to play the "base" role?

rtibbles commented:
`webpack.config.base.js` is the base config that gets used for all other webpack configs (so shared across tests, production building, and dev building).

`parse_bundle_plugin.js` is at the end of a chain of three functions that build the configuration for `webpack.config.js`, so acts as the point of entry for the base config to get into the webpack configuration.

I will annotate the relevant files to make their roles more clear. Thanks for the feedback!

indirectlylit commented on kolibri/core/templatetags/kolibri_tags.py:
```diff
@@ -36,15 +39,51 @@ def kolibri_main_navigation():
+    return frontend_sync(BASE_FRONTEND_SYNC)
+
+
+def frontend_async(hook):
```
I'm till having a hard time wrapping my head around the design and use of the "hook" enumerated constants.

For example, `def frontend_async(hook)` and `def frontend_sync(hook)` aren't used anywhere except in this file. Also, I can't find where the "callbacks" are defined, and what they return.

indirectlylit replied:
Ah, I see that subclasses can define a `hooks` method which returns a dict of functions. However these functions (e.g. `_register_front_end_plugins`, `main_navigation`) don't seem to have a consistent return value, but `get_webpack_bundle(bundle_path, extension)` seems to be called with the return value of the callbacks...

rtibbles replied:
Yeah - this is where each hook has an 'implicit' API - based on function that its callbacks eventually get passed to.

I agree this could be improved - but it was like that when I got here.

indirectlylit replied:
Above though we call `render_as_tags(get_webpack_bundle(callback(), None))` – doesn't that assume a particular return value from the callback?

rtibbles replied:
Yes, that's what I mean by the implicit API of the hook. `_register_front_end_plugins` should be the method that is passed to the `FRONTEND_PLUGIN` hook as a callback. It will break if any other method is passed.

indirectlylit commented on kolibri/core/templatetags/kolibri_tags.py:
```diff
@@ -36,15 +39,51 @@ def kolibri_main_navigation():
-        url = chunk.get('publicPath') or chunk['url']
         if chunk['name'].endswith('.js'):
-            tags.append('<script type="text/javascript" src="{static}/{url}"></script>'.format(static=static, url=url))
+            tags.append('<script type="text/javascript" src="{url}"></script>'.format(url=render_as_url(chunk)))
```
When module assets are loaded synchronously here (or asynchronously via `scriptjs`), it looks like they are executed in the global namespace. However files like `example_module.js` seem to leverage `module.exports` for namespacing. Is `module.exports` actually being used?

rtibbles replied:
After webpack bundling, they are anonymous - the `module.exports` in `example_module.js` is vestigial and could be removed.

indirectlylit replied:
@rtibbles clarified: the loaded JS is wrapped by `webpack` and not in the global namespace. However it is true that e.g. `module.exports = exampleModule;` has no effect, and all communication must happen through events with typical plugins rather than exported modules like in CommonJS and AMD.

rtibbles commented:
This PR is now a month old. It is also being used in active code development by @MCGallaspy and @DXCanas - I believe all interested parties have had the opportunity to review, and I have responded to all comments.

@aronasorman has said to me that he does not have the time or familiarity to give the go ahead, as such, and due to the foregoing that there has been ample review, I will be merging this at EOD today if there are no strenuous objections raised prior to then.

MCGallaspy commented:
Yeah, I think further iteration on this can be addressed in subsequent PRs.
